### DISCUSSION TITLE
Preview feature feedback: `@prisma/adapter-libsql` / Turso support

### DISCUSSION THREAD

**Discussion by janpio** [prisma/prisma#21345] [Feedback]
Support for **[Turso](https://turso.tech/)** is in early access behind the `driverAdapters` preview feature. You can learn about it in our [blog post](https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX).

Please share your feedback about `@prisma/adapter-libsql` released in [v5.4.0](https://github.com/prisma/prisma/releases/tag/5.4.0) in this issue.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.
- If you have any questions, don't hesitate to ask them in the [Prisma Slack](https://slack.prisma.io/) community

**Comment by jonstuebe** [comment]
I've updated to the latest package versions and I can't seem to add an `adapter` to my `PrismaClient`:

<img width="741" alt="CleanShot 2023-09-30 at 08 57 53@2x" src="https://github.com/prisma/prisma/assets/156722/0e98e819-0f01-485b-9a44-309f6088edfe">

**Reply by janpio** [reply]
Until Wednesday you have to use `@prisma/client@turso` - which points to a specific version. npm install `@prisma/client@turso` should fix that. (It seems we have to document that to make it more obvious). Does that work?

**Reply by Jolg42** [reply]
Now you can use `npm install @prisma/client` without `@turso` since `5.4.0` is released.

**Comment by edgarsilva** [comment]
@ruheni  @janpio first of all great work with the Turso libsql Adapter 👏🏽 !

 Is it a known issue that when using embedded replicas you can not write to the DB?

I was experimenting with this and embedded replicas as explained here -> [https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX](https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX) and found out all inserts fail with this error ->

![image](https://github.com/prisma/prisma/assets/518231/10ceade4-a1bb-454a-9517-987269933c54)

I'm wondering if this is really a Prisma issue or maybe a `libsql` problem, but you might have better insight.

```bash
[2] PrismaClientKnownRequestError: 
[2] Invalid `prisma.todo.create()` invocation:
[2] 
[2] 
[2] TRANSACTION_CLOSED: The transaction is closed
[2]  code: "TRANSACTION_CLOSED"
[2] 
[2]       at new U (/home/edgar/workspace/scaffold-apps/beth-stack/node_modules/@prisma/client/runtime/library.js:26:1479)
[2]       at handleRequestError (/home/edgar/workspace/scaffold-apps/beth-stack/node_modules/@prisma/client/runtime/library.js:123:6816)
[2]       at handleAndLogRequestError (/home/edgar/workspace/scaffold-apps/beth-stack/node_modules/@prisma/client/runtime/library.js:123:6200)
[2]       at /home/edgar/workspace/scaffold-apps/beth-stack/node_modules/@prisma/client/runtime/library.js:123:5920
[2]       at processTicksAndRejections (:61:76)
[2] POST - http://localhost:3000/api/todos failed
[2] 2023-11-05T14:27:30.140811Z ERROR tower_http::trace::on_failure: response failed classification=Code: 9 latency=42 ms
```

It seems the transaction is closed prematurely or something similar, going back to no `syncUrl` works great though, no issues with the adapter so far.

It might be worth mentioning somewhere that so far, due to this still being experimental, inserts to the db don't work while also using embedded replicas.

It's hard to go back though after experiencing the crazy fast query responses.

**Reply by janpio** [reply]
Yes, this is a Turso/LibSQL bug that snuck in: https://github.com/libsql/libsql-client-ts/issues/124
There is already a PR that will hopefully resolve this: https://github.com/tursodatabase/libsql/pull/559 
Just waiting for it to be merged and released, then updating to the last libsql driver should be enough.

**Reply by edgarsilva** [reply]
Great, thanks for the quick response, hopefully gets merged soon, hard to go back after you experience under 10ms queries  🏎️  🔥 ...

**Reply by rodolphoasb** [reply]
I think it's fixed now: https://github.com/tursodatabase/libsql/pull/577

**Reply by janpio** [reply]
Yes, but no release since merge yet: https://github.com/tursodatabase/libsql/releases

**Reply by janpio** [reply]
This should have been release in the meantime @edgarsilva. Can you maybe confirm?

**Reply by edgarsilva** [reply]
Nice, will test it over the weekend to make sure all CRUD operations work and update here 👍🏽

**Reply by edgarsilva** [reply]
@janpio confirmed is working 👍🏽 !

here are my current deps versions used to confirm it ->

Note: running on bun v1.0.25

```json
{
   "dependencies": {
    "@libsql/client": "^0.4.0-pre.2",
    "@prisma/adapter-libsql": "^5.5.2",
    "@prisma/client": "^5.5.2",
    "libsql": "^0.2.0-pre.7",
    "prisma": "^5.5.2",
  },
}
```

Read speeds are "blazingly fast" (less than 1ms locally), sadtly writes to the remote replicas are all over the place from around 400 ms the fastest to over 1s.

This is what I'm seeing on first sync (and every 30s my turso sync runs) ->

![Screenshot from 2024-01-21 15-21-32](https://github.com/prisma/prisma/assets/518231/75d78d4c-0f6a-43f0-96c3-b096c68dac8b)

A couple avg. writes (5th and 10th write in under 30s) ->
![Screenshot from 2024-01-21 15-24-03](https://github.com/prisma/prisma/assets/518231/650d34bd-28e2-48b2-b16e-2e2b72c4b142)

![Screenshot from 2024-01-21 15-45-23](https://github.com/prisma/prisma/assets/518231/3bbd3fed-1c8a-4f80-8f2c-b8032f921d87)


And this are the avg. read speeds from the local replicas.

![Screenshot from 2024-01-21 15-24-47](https://github.com/prisma/prisma/assets/518231/34d779be-5537-4ea9-9859-8742bcf806c1)

This is not good if you have a few writes, services like Supabase and PlanetScale easily give you under 200ms avg. per operation (of course depending on relative location) and way more stable times.

I'm wondering if too much sync after the write ant select that Prisma does could be hurting response times...

**Comment by Ivanrenes** [comment]
Really waiting for applying migrations with schema entity changes, is there anything in progress for that?

**Reply by janpio** [reply]
No, please leave a 👍 reaction on https://github.com/prisma/prisma/issues/22184 to register your interest.

**Comment by itsmylife** [comment]
Probably it is a known issue but just want to point this:

`@prisma/adapter-libsql@5.10.2` and `@libsql/client@0.5.3` is not compatible. 

```
❯ npm i
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: @prisma/adapter-libsql@5.10.2
npm ERR! Found: @libsql/client@0.5.3
npm ERR! node_modules/@libsql/client
npm ERR!   @libsql/client@"0.5.3" from the root project
```

**Reply by Jolg42** [reply]
@itsmylife Indeed, if you use `@prisma/adapter-libsql@early-access` or use `--force` or `--legacy-peer-deps` it would work.
Because it's technically compatible but the `peerDependencies` are on the conservative side, and because it's a `v0` package for `@libsql/client`, they need to be update for each minor version.

**Reply by itsmylife** [reply]
@Jolg42 Thank you for the reply. I'll do that. 
Apart from that, I'll keep getting build errors with nextjs@14.1.3, prisma@5.10.2 and libsql@0.5.3, nodejs@20.10.0
I don't think it is related to prisma but just want to share it to reach more people. Maybe somebody has an idea.

```
Failed to compile.

./node_modules/@libsql/darwin-arm64/README.md
Module parse failed: Unexpected character ' ' (1:1)
You may need an appropriate loader to handle this file type, currently, no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> # `@libsql/darwin-arm64`
| 
| Prebuilt binary package for `libsql` on `darwin-arm64`.
```

**Reply by Jolg42** [reply]
@itsmylife I found a related issue on Turso's repo https://github.com/tursodatabase/libsql-client-ts/issues/184, maybe you can post there.

**Reply by Jolg42** [reply]
I think you could try to manually add the package to `serverComponentsExternalPackages`
see https://github.com/vercel/next.js/blob/canary/docs/02-app/02-api-reference/05-next-config-js/serverComponentsExternalPackages.mdx)

**Comment by Hacksore** [comment]
Is this still working?

![image](https://github.com/user-attachments/assets/fea188ae-edb8-40f8-a3d9-8afee6be1dc0)

seeing this error:
```
Error: Cannot read properties of undefined (reading 'bind')
```

issue we are using for tracking the db work.
https://github.com/Hacksore/oghunt/issues/48

Code for those who wanna take a look.
https://github.com/Hacksore/oghunt/tree/db

**Reply by Hacksore** [reply]
We actually got it working but it seems way to slow as of now 😭

**Comment by touhidurrr** [comment]
https://github.com/prisma/prisma/discussions/25931

Basically I can find no way to use Turso without embedded replicas.

**Comment by kojikokojiko** [comment]
Hi there,

I've encountered an issue while using the Prisma Client extension for turso DB.

When executing the following INSERT statement:

```
const feed = await prisma.feed.create({
  data: {
    url: body.url,
    name: body.name,
    description: body.description || '',
    imageUrl: body.imageUrl || ''
  }
});
I'm
```

 receiving this error message:

```
⨯ TypeError: The "payload" argument must be of type object. Received null
at POST
```
It seems like there's an issue with the return value of the create operation. Could you please advise on how to resolve this issue?

Thank you in advance for your assistance!

Best regards,
koji.iwase

**Comment by fermeridamagni** [comment]
Hi there

I've found a issue while using the Prisma Client with the SQLite Adapter (Turso).

- Enviroiment
  - Windows 11

```package.json

"next": "15.2.2",
"react": "^19.0.0",
"react-dom": "^19.0.0",
```

- `prisma` & `libSQL` dependencies:
```package.json

 "@libsql/client": "^0.14.0",
 "@prisma/adapter-libsql": "^6.5.0",
 "@prisma/client": "^6.5.0",
 "prisma": "^6.5.0",
```

When i executing the `npm i` command it throws an error like this.

```cmd
npm i
npm error code ERESOLVE
npm error ERESOLVE could not resolve
npm error
npm error While resolving: @prisma/adapter-libsql@6.5.0
npm error Found: @libsql/client@0.14.0
npm error node_modules/@libsql/client
npm error   @libsql/client@"^0.14.0" from the root project
npm error
npm error Could not resolve dependency:
npm error peer @libsql/client@"^0.3.5 || ^0.4.0 || ^0.5.0 || ^0.6.0 || ^0.7.0 || ^0.8.0" from @prisma/adapter-libsql@6.5.0
npm error node_modules/@prisma/adapter-libsql
npm error   @prisma/adapter-libsql@"^6.5.0" from the root project
npm error
npm error Conflicting peer dependency: @libsql/client@0.8.1
npm error node_modules/@libsql/client
npm error   peer @libsql/client@"^0.3.5 || ^0.4.0 || ^0.5.0 || ^0.6.0 || ^0.7.0 || ^0.8.0" from @prisma/adapter-libsql@6.5.0
npm error   node_modules/@prisma/adapter-libsql
npm error     @prisma/adapter-libsql@"^6.5.0" from the root project
npm error
npm error Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
```

I've been working correctly because if i use the `--legacy-peer-deps` it works.

There are any fix for this?

Best regards,
Fer Mérida

**Comment by jpcafe10** [comment]
Hi all, tried out the early access for `db push`. 
Getting this "error: Error validating datasource `db`: the URL must start with the protocol `file:`."

Versions: 
```
"@libsql/client": "^0.15.2",
"@prisma/adapter-libsql": "^6.6.0",
"prisma": "^6.6.0",
```

my config file

```ts
import path from 'node:path';
import { defineConfig } from 'prisma/config';
import { PrismaLibSQL } from '@prisma/adapter-libsql';

type Env = {
	DATABASE_URL: string;
	DATABASE_AUTH_TOKEN: string;
};

const migrateConfig = process.env.UNSUPPORTED_PRISMA_MIGRATION
	? {}
	: {
			migrate: {
				async adapter(env) {
					return new PrismaLibSQL({
						url: env.DATABASE_URL,
						authToken: env.DATABASE_AUTH_TOKEN
					});
				}
			}
		};

export default defineConfig<Env>({
	earlyAccess: true,
	schema: path.join('prisma', 'schema.prisma'),
	studio: {
		async adapter(env) {
			return new PrismaLibSQL({
				url: env.DATABASE_URL,
				authToken: env.DATABASE_AUTH_TOKEN
			});
		}
	},
	...migrateConfig
});
```

**Reply by touhidurrr** [reply]
You can give it a random file path (maybe something like /temp/test.db) and it will work. Migrations will be done to the local db, not the turso one if I remember correctly. Do double check though.

Note: The actually SQLs would run on the main database. Only migrations are different. Again not sure, it has been a while since I did it. So double check before proceeding.

**Reply by jpcafe10** [reply]
according to the docs https://www.prisma.io/docs/orm/overview/databases/turso#2-set-environment-variables
`db push` is supported

**Reply by jpcafe10** [reply]
oh you're suggesting using a different env variable for the actual DATABASE_URL in schema.prisma... i will try

**Reply by touhidurrr** [reply]
No. Just put a useless file path in prisma schema and the main database url in PrismaLibSQL constructor.

**Reply by jpcafe10** [reply]
yeah i got it thanks!

**Reply by jpcafe10** [reply]
nice it worked, thanks a lot 😄 
Suggestion would be to maybe change this entry in the cli:
```
Prisma schema loaded from prisma/schema.prisma
Datasource "db": SQLite database "dev.db" at "file:../db/dev.db"

The database is already in sync with the Prisma schema.
```

**Comment by Arlen22** [comment]
https://github.com/prisma/prisma/issues/26888

In case anyone else runs into this in 6.6.0.

**Comment by Arlen22** [comment]
I'm trying to use the turso database adapter to get end-to-end wasm with prisma, but I'm not sure what all the pieces are that I need to put together. Is this even supported at the moment? Do I need to write my own driver adapter?
